### PR TITLE
gh-110812: Isolating Extension Modules HOWTO: List GC-related gotchas

### PR DESCRIPTION
**Pull request by encukou**
Here's more text on how to avoid the GC-related gotchas when defining heap types.

(See also: https://github.com/capi-workgroup/problems/issues/49)

* Issue: gh-110812

----
:books: Documentation preview :books:: https://cpython-previews--111504.org.readthedocs.build/

### CODE REVIEW

File: Doc/howto/isolating-extensions.rst [hunk 1]
```diff
@@ -339,12 +339,44 @@ That is, heap types should:
...
+       }
+       return 0;
+   }
+
+Unfortunately, :c:data:`Py_Version` was only added in Python 3.11.
+As a replacement, use:
+
+* :c:macro:`PY_VERSION_HEX`, if not using the stable ABI, or
+* :py:data:`sys.version_info` (via :c:func:`PySys_GetObject` and
+  :c:func:`PyArg_ParseTuple`).
```
Comment (ericsnowcurrently) [inline Doc/howto/isolating-extensions.rst:371]: Consider providing a second example that works on 3.9 and 3.10.  Being able to copy-and-paste in a situation like this take the sting away a little.

Comment (ericsnowcurrently) [reply Doc/howto/isolating-extensions.rst:371]: or a palatable example that works for 3.9+.

File: Doc/howto/isolating-extensions.rst [hunk 2]
```diff
@@ -356,11 +388,59 @@ For example, if your traverse function includes::
...
+if your type does *not* override
+``tp_dealloc`` you don't need to add it.
+
+
+Not overriding ``tp_free``
+..........................
+
+The :c:member:`~PyTypeObject.tp_free` slot of a heap type must be set to
+:c:func:`PyObject_GC_Del`.
+This is the default; avoid overriding it.
```
Comment (ericsnowcurrently) [inline Doc/howto/isolating-extensions.rst]: Are there situations in which they might need to?  If not, I recommend using firmer language:

```suggestion
This is the default; do not override it.
```

If there *is* a valid reason then consider providing brief instructions and perhaps a short example (after strongly encouraging folks not to do it).  Again, we want to reduce friction as much as possible for folks that are trying to do the right thing by switching to heap types.  One extra example here might be enough to keep someone from going past their frustration threshold.  (I know I'm preaching to the choir, @encukou, but I want to be extra clear for anyone else the comes along. 😄)

File: Doc/howto/isolating-extensions.rst [hunk 3]
```diff
@@ -356,11 +388,59 @@ For example, if your traverse function includes::
...
+Avoiding ``PyObject_New``
+.........................
+
+GC-tracked objects need to be allocated using GC-aware functions.
+
+If you use use :c:func:`PyObject_New` or :c:func:`PyObject_NewVar`:
+
+- Get and call type's :c:member:`~PyTypeObject.tp_alloc` slot , if possible.
+- If not possible (e.g. inside a custom ``tp_alloc``),
+  call :c:func:`PyObject_GC_New` or :c:func:`PyObject_GC_NewVar`.
```
Comment (ericsnowcurrently) [inline Doc/howto/isolating-extensions.rst]: A copy-and-pasteable example would be helpful.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Thanks for following up on this, @encukou!  I think this is a nice improvement, though I'll rely on @1st1 to say for sure. 😄  I've left a few small comments.

**Comment by ericsnowcurrently** [comment]
@pablogsal, do you think this change would have helped Yury get things right without your help?

File: Doc/howto/isolating-extensions.rst (hunk 2 above)
Comment (encukou) [reply Doc/howto/isolating-extensions.rst]: > Are there situations in which they might need to?

I don't know of any.

File: Doc/howto/isolating-extensions.rst (hunk 1 above)
Comment (encukou) [reply Doc/howto/isolating-extensions.rst:371]: Let's leave that to a follow-up PR?

**Comment by pablogsal** [comment]
> @pablogsal, do you think this change would have helped Yury get things right without your help?

Yup! This is indeed fantastic. Thank you @ericsnowcurrently and @encukou

File: Doc/howto/isolating-extensions.rst [hunk 4]
```diff
@@ -438,9 +438,20 @@ GC-tracked objects need to be allocated using GC-aware functions.
 
 If you use use :c:func:`PyObject_New` or :c:func:`PyObject_NewVar`:
 
-- Get and call type's :c:member:`~PyTypeObject.tp_alloc` slot , if possible.
-- If not possible (e.g. inside a custom ``tp_alloc``),
-  call :c:func:`PyObject_GC_New` or :c:func:`PyObject_GC_NewVar`.
+- Get and call type's :c:member:`~PyTypeObject.tp_alloc` slot, if possible.
+  That is, replace ``TYPE *o = PyObject_New(TYPE, typeobj)`` by::
```
Comment (ericsnowcurrently) [inline Doc/howto/isolating-extensions.rst]:
```suggestion
  That is, replace ``TYPE *o = PyObject_New(TYPE, typeobj)`` with::
```

File: Doc/howto/isolating-extensions.rst [hunk 5]
```diff
@@ -438,9 +438,20 @@ GC-tracked objects need to be allocated using GC-aware functions.
...
+
+      TYPE *o = typeobj->tp_alloc(typeobj, 0);
```
Comment (ericsnowcurrently) [inline Doc/howto/isolating-extensions.rst:444]: It's a shame `PyObject_New()` (and `PyObject_NewVar()`) don't take care of this.  Assuming they can't, would it make sense to provide a new function that does the right thing?  That way we avoid people relying on `tp_alloc`.

<details>
<summary>For example:</summary>

```C
// Include/object.c
PyAPI_FUNC(PyObject *) _PyObject_NewEx(PyTypeObject *);
PyAPI_FUNC(PyObject *) _PyObject_NewVarEx(PyTypeObject *, Py_ssize_t);

#define PyObject_NewEx(type, typeobj) \
    ((type *)_PyObject_NewEx(typeobj))
#define PyObject_NewVarEx(type, typeobj, n) \
    ((type *)_PyObject_NewVarEx(typeobj, n))
```

```C
// Objects/object.c
PyObject *
_PyObject_NewEx(PyTypeObject *tp)
{
    return tp->tp_alloc(tp, 0);
}

PyObject *
_PyObject_NewVarEx(PyTypeObject *tp, Py_ssize_t nitems)
{
    return tp->tp_alloc(tp, nitems);
}
```

or even:

```C
// Objects/object.c
PyObject *
_PyObject_NewEx(PyTypeObject *tp)
{
    if (_PyType_IS_GC(tp) && tp->tp_alloc == PyBaseObject.tp_alloc) {
        return _PyObject_GC_New(tp);
    }
    return tp->tp_alloc(tp, 0);
}

PyObject *
_PyObject_NewVarEx(PyTypeObject *tp, Py_ssize_t nitems)
{
    if (_PyType_IS_GC(tp) && tp->tp_alloc == PyBaseObject.tp_alloc) {
        return _PyObject_GC_NewVar(tp, nitems);
    }
    return tp->tp_alloc(tp, nitems);
}
```

</details>

Comment (encukou) [reply Doc/howto/isolating-extensions.rst:444]: There are many things that could be improved with new API. I'd like to keep this PR focused on the HOWTO, which will need to cover old versions anyway.

More generally, I think it'd be good to provide “caller” API for all the slots. Something like that will be necessary if we ever want to start changing the slot signatures.

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

**Pull request merged**